File: contacting-conf-web/src/main/java/nl/ing/api/contacting/conf/service/BlacklistService.java
package nl.ing.api.contacting.conf.service;

import com.ing.api.contacting.dto.java.context.ContactingContext;
import com.ing.api.contacting.dto.java.resource.blacklist.BlacklistFunctionality;
import com.ing.api.contacting.dto.java.resource.blacklist.BlacklistItemDto;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.ing.api.contacting.conf.domain.entity.BlacklistEntity;
import nl.ing.api.contacting.conf.domain.model.blacklist.BlacklistItemVO;
import nl.ing.api.contacting.conf.exception.Errors;
import nl.ing.api.contacting.conf.mapper.BlacklistItemMapperJava;
import nl.ing.api.contacting.conf.repository.BlacklistJpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class BlacklistService {

    private final BlacklistJpaRepository repository;

    @Transactional(readOnly = true)
    public List<BlacklistItemDto> getAllBlacklistItems(ContactingContext contactingContext) {
        return toDtoList(repository.findActiveByAccount(contactingContext.accountId(), LocalDateTime.now()));
    }

    @Transactional(readOnly = true)
    public List<BlacklistItemDto> getAllByFunctionality(String functionality, ContactingContext context) {

        BlacklistFunctionality func = BlacklistFunctionality.fromString(functionality)
                .orElseThrow(() -> Errors.notFound(functionality + " is not defined"));

        LocalDateTime now = LocalDateTime.now();
        List<BlacklistEntity> entities = (func == BlacklistFunctionality.ALL)
                ? repository.findActiveByAccount(context.accountId(), now)
                : repository.findByAccountIdAndFunctionalityAndActive(context.accountId(), func.value(), now);
        return toDtoList(entities);
    }

    private List<BlacklistItemDto> toDtoList(List<BlacklistEntity> entities) {
        return entities.stream()
                .map(BlacklistItemMapperJava::toVO)
                .map(BlacklistItemMapperJava::toDto)
                .toList();
    }


    @Transactional
    public BlacklistItemVO createBlackListItem(BlacklistItemVO vo, ContactingContext contactingContext) {
        BlacklistEntity entityToSave = BlacklistItemMapperJava.toEntity(vo, contactingContext.accountId());
        BlacklistEntity saved = repository.save(entityToSave);
        return BlacklistItemMapperJava.toVO(saved);
    }

    @Transactional
    public BlacklistItemVO updateBlackListItem(BlacklistItemVO vo, ContactingContext contactingContext) {
        BlacklistEntity entity = BlacklistItemMapperJava.toEntity(vo, contactingContext.accountId());
        BlacklistEntity updated = repository.save(entity);
        return BlacklistItemMapperJava.toVO(updated);
    }

    @Transactional
    public void deleteBlackListItem(Long id) {
        repository.deleteById(id);
    }
}

File: contacting-conf-web/src/main/scala/nl/ing/api/contacting/conf/business/BlacklistService.scala
package nl.ing.api.contacting.conf.business

import cats.effect.{Async, IO}
import com.ing.api.contacting.dto.context.ContactingContext
import com.ing.api.contacting.dto.resource.account.AccountDto
import com.ing.api.contacting.dto.resource.blacklist.BlacklistFunctionality
import com.ing.api.contacting.dto.resource.blacklist.BlacklistFunctionality.BlacklistFunctionality
import nl.ing.api.contacting.conf.repository.BlacklistRepository
import nl.ing.api.contacting.conf.repository.cslick.data.BlacklistItemData
import nl.ing.api.contacting.conf.util.ImplicitLazyLogging
import nl.ing.api.contacting.tracing.Trace

import java.time.LocalDateTime

/**
  * BlacklistService
  *
  * Provides some functionality to handle the basic CRUD operations on the blacklist.
  */
class BlacklistService[F[_]: Async : Trace](blacklistRepository: BlacklistRepository[F])
    extends ImplicitLazyLogging {

  def getAllBlacklistItemsByFunctionality(
      blacklistFunctionality: BlacklistFunctionality,
      context: ContactingContext): F[Seq[BlacklistItemData]] =
    blacklistFunctionality match {
      case BlacklistFunctionality.ALL => this.getAllBlacklistItems(context)
      case _ =>
        blacklistRepository
          .getAllBlacklistItemsByFunctionality(context.accountId,
                                               LocalDateTime.now,
                                               blacklistFunctionality)
    }

  def getAllBlacklistItems(
      context: ContactingContext): F[Seq[BlacklistItemData]] = {
    blacklistRepository
      .getAllBlacklistItems(context.accountId, LocalDateTime.now)
  }

  def getAllBlacklistItems(accountSid: String): F[Seq[BlacklistItemData]] = {
    blacklistRepository
      .getAllSurveyBlacklistItems(accountSid, LocalDateTime.now)
  }

  def cleanUpBlacklist(implicit account: AccountDto): F[Int] = {
    blacklistRepository.cleanUpBlacklist(account.id)
  }

  def createBlackListItem(entity: BlacklistItemData,
                          context: ContactingContext): F[BlacklistItemData] = {
    blacklistRepository
      .save(entity, context)
  }

  def updateBlackListItem(entity: BlacklistItemData,
                          context: ContactingContext): F[BlacklistItemData] = {
    blacklistRepository
      .update(entity, context)
  }

  def deleteBlackListItem(id: Long, context: ContactingContext): F[Int] = {
    blacklistRepository
      .deleteById(id, context)
  }
}
